<launch>

    <!-- Choose visualization -->
    <arg name="rtabmap" default="true"/>
    <arg name="rtabmapviz" default="false"/>

    <arg if="$(arg rtabmapviz)" name="rviz" default="false"/>
    <arg unless="$(arg rtabmapviz)" name="rviz" default="true"/>

    <!-- Localization-only mode -->
    <arg name="localization" default="false"/>
    <arg if="$(arg localization)" name="deleteDB" default="false"/>
    <arg unless="$(arg localization)" name="deleteDB" default="true"/>

    <arg name="database_path" default="~/.ros/rtabmap.db"/>

    <!-- Corresponding config files -->
    <arg name="cfg" default=""/>
    <!-- To change RTAB-Map's parameters, set the path of config file (*.ini) generated by the standalone app -->
    <arg name="gui_cfg" default="~/.ros/rtabmap_gui.ini"/>
    <arg name="rviz_cfg" default="$(find automap)/launch/config/loomo.rviz"/>

    <!-- Args / output -->
    <arg if="$(arg deleteDB)" name="args" default="--delete_db_on_start"/>
    <!-- delete_db_on_start, udebug -->
    <arg unless="$(arg deleteDB)" name="args" default=""/>
    <arg name="output" default="screen"/>
    <!-- Control node output (screen or log) -->

    <!-- Image Topics -->
    <arg name="rgb_topic" default="/camera/rgb/image_raw"/>
    <arg name="rgb_info_topic" default="/camera/rgb/camera_info"/>
    <arg name="depth_topic" default="/camera/depth/image_raw"/>
    <arg name="depth_info_topic" default="/camera/depth/camera_info"/>

    <!-- Other Topics -->
    <arg name="odom_topic" default="/odom"/>
    <arg name="odom_frame_id" default="odom"/>

    <!-- Frame id's -->
    <arg name="frame_id" default="base_link"/>
    <arg name="map_frame_id" default="map"/>

    <!-- RGBD -->
    <arg name="subscribe_rgbd" default="true"/>
    <arg name="rgbd_topic" default="/camera/rgbd/image_raw"/>

    <!-- Scan -->
    <arg name="subscribe_scan" default="false"/>
    <arg name="scan_topic" default="/scan_filtered"/>

    <arg name="publish_tf_map" default="true"/>

    <arg if="$(arg subscribe_rgbd)" name="subscribe_depth" default="false"/>
    <arg unless="$(arg subscribe_rgbd)" name="subscribe_depth" default="true"/>

    <arg if="$(arg subscribe_rgbd)" name="subscribe_rgb" default="false"/>
    <arg unless="$(arg subscribe_rgbd)" name="subscribe_rgb" default="true"/>


    <arg if="$(arg localization)" name="incremental_memory" default="false"/>
    <arg unless="$(arg localization)" name="incremental_memory" default="true"/>

    <arg name="subscribe_user_data" default="false"/>
    <!-- user data synchronized subscription -->
    <arg name="user_data_topic" default="/marker_signal"/>
    <arg name="user_data_async_topic" default="/marker_signal"/>
    <!-- user data async subscription (rate should be lower than map update rate) -->

    <arg name="approx_sync" default="true"/>
    <arg name="queue_size" default="20"/>

    <!-- Nodes -->

    <!-- Simultaneous Localization And Mapping -->
    <node if="$(arg rtabmap)" name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="$(arg output)" args="$(arg args)">

        <param name="database_path" value="$(arg database_path)"/>

        <remap from="rgbd_image" to="$(arg rgbd_topic)"/>
        <remap from="odom" to="$(arg odom_topic)"/>
        <remap from="scan" to="$(arg scan_topic)"/>
        <remap from="user_data_async" to="$(arg user_data_topic)"/>
        <!-- <remap from="user_data" to="$(arg user_data_topic)" /> -->

        <param name="frame_id" type="string" value="$(arg frame_id)"/>

        <param name="subscribe_rgb" type="bool" value="$(arg subscribe_rgb)"/>
        <param name="subscribe_rgbd" type="bool" value="$(arg subscribe_rgbd)"/>
        <param name="subscribe_depth" type="bool" value="$(arg subscribe_depth)"/>
        <param name="subscribe_scan" type="bool" value="$(arg subscribe_scan)"/>
        <param name="subscribe_user_data" type="bool" value="$(arg subscribe_user_data)"/>

        <param name="approx_sync" type="bool" value="$(arg approx_sync)"/>
        <param name="config_path" type="string" value="$(arg cfg)"/>
        <param name="queue_size" type="int" value="$(arg queue_size)"/>

        <param name="map_always_update" type="bool" value="false"/>

        <!-- RTABMap's parameters -->
        <param name="RGBD/ProximityBySpace" type="string" value="false"/>
        <param name="RGBD/AngularUpdate" type="string" value="0.1"/>
        <param name="RGBD/LinearUpdate" type="string" value="0.1"/>
        <param name="RGBD/OptimizeMaxError" type="string" value="0.0"/>
        <param name="RGBD/OptimizeRobust" type="string" value="true"/>

        <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>

        <param name="Optimizer/Strategy" type="string" value="2"/>
        <!-- g2o=1, GTSAM=2 -->
        <param name="Optimizer/Robust" type="string" value="true"/>
        <param name="Optimizer/MaxError" type="string" value="0"/>

        <!-- Occupancy Grid -->
        <param name="Grid/3D" type="bool" value="true"/>
        <param name="Grid/FromDepth" type="bool" value="true"/>
        <param name="Grid/MaxObstacleHeight" type="double" value="1.8"/>
        <param name="Grid/Scan2dUnknownSpaceFilled" type="bool" value="true"/>
        <param name="Grid/RangeMax" type="double" value="3.5"/>
        <param name="Grid/NoiseFilteringRadius" type="double" value="0"/>
        <param name="Grid/NoiseFilteringMinNeighbors" type="int" value="5"/>
        <param name="Grid/CellSize" type="double" value="0.05"/>
        <param name="Grid/RayTracing" type="bool" value="true"/>

        <param name="Reg/Force3DoF" type="string" value="true"/>

        <param name="Vis/MinInliers" type="string" value="10"/>
        <param name="Vis/InlierDistance" type="string" value="0.1"/>

        <param name="Rtabmap/DetectionRate" type="string" value="2"/>
        <param name="Rtabmap/TimeThr" type="string" value="350"/>

        <!-- Localization mode -->
        <param name="Mem/IncrementalMemory" type="string" value="$(arg incremental_memory)"/>
        <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>

    </node>

    <!-- Visualisation RTAB-Map -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(arg gui_cfg)" output="$(arg output)">

        <remap from="rgbd_image" to="$(arg rgbd_topic)"/>
        <remap from="scan" to="$(arg scan_topic)"/>
        <remap from="odom" to="$(arg odom_topic)"/>

        <param name="subscribe_rgbd" type="bool" value="$(arg subscribe_rgbd)"/>
        <param name="subscribe_depth" type="bool" value="$(arg subscribe_depth)"/>
        <param name="subscribe_scan" type="bool" value="$(arg subscribe_scan)"/>

        <param name="frame_id" type="string" value="$(arg frame_id)"/>

        <param name="queue_size" type="int" value="$(arg queue_size)"/>
        <param name="approx_sync" type="bool" value="$(arg approx_sync)"/>

        <param name="wait_for_transform_duration" type="double" value="0.2"/>

    </node>

    <!-- Visualization RVIZ -->
    <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(arg rviz_cfg)"/>

</launch>
